pkg/pool-weighted/contracts/WeightedMath.sol

```diff
@@ -407,41 +407,58 @@ library WeightedMath {
-        uint256 totalSupply,
-        uint256 previousInvariant,
-        uint256 currentInvariant,
+        uint256 invariantGrowthRatio,
+        uint256 previousSupply,
+        uint256 currentSupply,
         uint256 protocolSwapFeePercentage
-    ) internal pure returns (uint256) {
-        // We round down to prevent issues in the Pool's accounting, even if it means paying slightly less in protocol
-        // fees to the Vault.
-        uint256 growth = currentInvariant.divDown(previousInvariant);
-
-        // Shortcut in case there was no growth when comparing the current against the previous invariant.
-        // This shouldn't happen outside of rounding errors, but have this safeguard nonetheless to prevent the Pool
-        // from entering a locked state in which joins and exits revert while computing accumulated swap fees.
-        if (growth <= FixedPoint.ONE) {
-            return 0;
-        }
+    ) internal pure returns (uint256 protocolFeeAmount) {
+        uint256 supplyGrowthRatio = currentSupply.divDown(previousSupply);
+
+        // Joins and exits are symmetrical; for simplicity, we consider a join, where the invariant and supply
+        // both increase.
+
+        // |-------------------------|-- original invariant * invariantGrowthRatio
+        // |   increase from fees    |
+        // |-------------------------|-- original invariant * supply growth ratio (fee-less invariant)
+        // |                         |
+        // | increase from balances  |
+        // |-------------------------|-- original invariant
+        // |                         |
+        // |                         |  |------------------|-- currentSupply
+        // |                         |  |    BPT minted    |
+        // |                         |  |------------------|-- previousSupply
+        // |   original invariant    |  |  original supply |
+        // |_________________________|  |__________________|
+        //
+        // If the join is proportional, the invariant and supply will likewise increase proportionally,
+        // so the growth ratios (invariantGrowthRatio / supplyGrowthRatio) will be equal. In this case, we do not charge
+        // any protocol fees.
 
-        // Assuming the Pool is balanced and token weights have not changed, a growth of the invariant translates into
-        // proportional growth of all token balances. The protocol is due a percentage of that growth: more precisely,
-        // it is due `k = protocol fee * (growth - 1) * balance / growth` for each token.
-        // We compute the amount of BPT to mint for the protocol that would allow it to proportionally exit the Pool and
-        // receive these balances. Note that the total BPT supply will increase when minting, so we need to account for
-        // this in order to compute the percentage of Pool ownership the protocol will have.
+        if (supplyGrowthRatio == invariantGrowthRatio || protocolSwapFeePercentage == 0) return 0;
```

Comment:
```suggestion
        if ((supplyGrowthRatio == invariantGrowthRatio) || (protocolSwapFeePercentage == 0)) return 0;
```

plz no operator precedence abuse

Reply:
ser, there's no other way to operate these operators.